src/librustc_resolve/late/diagnostics.rs

```diff
@@ -253,6 +253,37 @@ impl<'a> LateResolutionVisitor<'a, '_> {
                 }
                 return (err, candidates);
             }
+
+            // Check if the first argument is `self` and suggest calling a method.
+            let mut has_self_arg = false;
+            if let PathSource::Expr(parent) = source {
+                match &parent.map(|p| &p.kind) {
+                    Some(ExprKind::Call(_, args)) if args.len() > 0 => {
+                        let mut expr_kind = &args.first().unwrap().kind;
+                        loop {
+                            match expr_kind {
+                                ExprKind::Path(_, arg_name) if arg_name.segments.len() == 1 => {
+                                    has_self_arg = arg_name.segments[0].ident.name == kw::SelfLower;
+                                    break;
+                                },
+                                ExprKind::AddrOf(_, _, expr) => { expr_kind = &expr.kind; }
+                                _ => break,
+                            }
+                        }
+                    }
+                    _ => (),
+                }
+            };
+
+            if has_self_arg {
+                err.span_suggestion(
+                    span,
+                    &"try calling method instead of passing `self` as parameter",
+                    format!("self.{}", path_str),
```

Comment:
This doesn't get rid of the existing `self`/`&self`, etc. at the moment.

Reply:
Could you elaborate? I'm not sure if I understand this.

Reply:
This was really the same comment as https://github.com/rust-lang/rust/pull/66913#discussion_r355130743.